### PR TITLE
Basic implementation of line decoration display items.

### PR DESCRIPTION
**Pull request by glennw**
Includes the commits in https://github.com/servo/webrender/pull/1491 from @Gankro.

This provides a basic implementation of the line decoration styles in CSS.

I haven't specifically tried to make the styles (e.g. wavy, dotted) match other browsers, yet. I think it's reasonable to get this reviewed and merged, and then tweak the visuals as a follow up.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/1504)

### CODE REVIEW

**Comment by glennw** [comment]
Sample of the current line styles, including text-shadow blur support:

decorations-suite

File: webrender/src/prim_store.rs [hunk 1]
```diff
@@ -180,6 +181,24 @@ impl ToGpuBlocks for RectanglePrimitive {
...
+pub struct LinePrimitive {
+    pub color: ColorF,
+    pub style: LineStyle,
+    pub orientation: LineOrientation,
+}
+
+impl ToGpuBlocks for LinePrimitive {
+    fn write_gpu_blocks(&self, mut request: GpuDataRequest) {
+        request.push(self.color);
+        request.push([pack_as_float(self.style as u32),
```
Comment (kvark) [inline webrender/src/prim_store.rs]: what's the reason to call `pack_as_float` here as opposed to just casting to `f32`?

File: webrender/src/renderer.rs [hunk 2]
```diff
@@ -1878,6 +1904,22 @@ impl Renderer {
                                       &target.text_run_textures,
                                       &projection);
         }
+        if !target.line_cache_prims.is_empty() {
+            // TODO(gw): Technically, we don't need blend for solid
+            //           lines. We could check that here?
```
Comment (kvark) [inline webrender/src/renderer.rs]: if that becomes a performance concern, we'd need to split those into opaque/non-opaque

**Comment by Gankro** [comment]
Seems good!

I kinda lost track of what you really wanted with the line-dimension specifying stuff. From servo's perspective, boxes + orientation would be easier to send (because of how their logical->real conversion works). Gecko would probably be fine either way?

**Comment by Gankro** [comment]
Also it appears that you're trying to make waves have a fixed width, and you just stretch the height. Gecko's algorithm is to have a fixed slope (45 degrees), so the width of a wave increases with its height:

screen shot 2017-07-20 at 3 46 51 pm 1

File: webrender/res/ps_line.fs.glsl [hunk 3]
```diff
@@ -0,0 +1,123 @@
+#line 1
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+float det(vec2 a, vec2 b) {
```
Comment (glennw) [inline webrender/res/ps_line.fs.glsl]: It's a 2d cross product, I guess. It's just using the same name as the paper the algorithm was taken from.

File: webrender/src/prim_store.rs (hunk 1 above)
Comment (glennw) [inline webrender/src/prim_store.rs]: ```pack_as_float``` adds 0.5 to ensure that casting back to an int definitely gets the correct value. It's more a defensive measure than anything - perhaps it's not necessary?

File: webrender/src/renderer.rs (hunk 2 above)
Comment (glennw) [inline webrender/src/renderer.rs]: Yup

**Pull request merged**